mesonbuild/build.py

```diff
@@ -1067,15 +1067,23 @@ def extract_all_objects(self, recursive: bool = True) -> ExtractedObjects:
         return ExtractedObjects(self, self.sources, self.generated, self.objects,
                                 recursive, pch=True)
 
+    def get_toplevel_link_deps(self) -> ImmutableListProtocol[BuildTargetTypes]:
+        return self.link_targets
+
     def get_all_link_deps(self) -> ImmutableListProtocol[BuildTargetTypes]:
         return self.get_transitive_link_deps()
 
     @lru_cache(maxsize=None)
     def get_transitive_link_deps(self) -> ImmutableListProtocol[BuildTargetTypes]:
-        result: T.List[Target] = []
-        for i in self.link_targets:
-            result += i.get_all_link_deps()
-        return result
+        result: T.Set[Target] = []
+        stack: T.Deque[Target] = deque()
+        stack.appendleft(self)
```

Comment:
You should add self only if it's a SharedLibrary

Reply:
Self is the root of the stack - it never actually gets added to `result`.

Reply:
I think you need to add self.get_toplevel_link_deps() to the stack initially.

Reply:
Not necessary. When the loop is first run, the only thing in the stack is `self` and `result` is empty, so `stack` gets filled with `self.get_toplevel_link_deps()` anyway. It could be changed to skip that first step, but that would make the function more complicated for little gain.